pcmaudiotrack.go

```diff
@@ -0,0 +1,197 @@
+	})
+
+	if isFirstSample {
+		return err
+	}
+
+	t.mu.Lock()
+	t.sampleBuffer = append(t.sampleBuffer, sample)
+	t.cond.Broadcast()
+	t.mu.Unlock()
+	return nil
+}
+
+func (t *PCM16ToOpusAudioTrack) processSamples() {
+	for range t.ticker.C {
+		t.mu.Lock()
+		var resetTimer bool
+
+		for len(t.sampleBuffer) == 0 && !t.closed.Load() {
+			t.cond.Wait()
+			resetTimer = true
+		}
+		if t.closed.Load() {
+			t.mu.Unlock()
+			return
+		}
+
+		sample := t.sampleBuffer[0]
+		if len(t.sampleBuffer) > 1 {
+			t.sampleBuffer = t.sampleBuffer[1:]
+		} else {
+			t.sampleBuffer = make([]audio.PCM16Sample, 0)
+		}
+
+		if resetTimer {
+			t.ticker.Reset(t.frameDuration)
+		}
+
+		t.resampledPCMWriter.WriteSample(sample)
+		t.mu.Unlock()
+	}
+}
+
+func (t *PCM16ToOpusAudioTrack) Close() {
+	t.closed.Store(true)
+	t.cond.Broadcast()
+	t.ticker.Stop()
+	t.resampledPCMWriter.Close()
+	t.pcmWriter.Close()
+	t.opusWriter.Close()
+}
+
+type OpusToPCM16AudioTrack struct {
+	*webrtc.TrackRemote
+	channels   int
+	sampleRate int
+
+	opusWriter audio.WriteCloser[opus.Sample]
+	pcmMWriter audio.WriteCloser[audio.PCM16Sample]
+	logger     protoLogger.Logger
+}
+
+// TODO: We also have a reader API, but writer is more efficient as it is zero copy.
+// Shall we support both reader and writer?
+// Reader makes more sense while reading the code, might be easier for the end user to understand.
+// But, it's less efficient as it involves a copy.
+func NewOpusToPCM16AudioTrack(track *webrtc.TrackRemote, publication *RemoteTrackPublication, writer *audio.WriteCloser[audio.PCM16Sample], sampleRate int, handleJitter bool) (*OpusToPCM16AudioTrack, error) {
+	if track.Codec().MimeType != webrtc.MimeTypeOpus {
+		return nil, errors.New("track is not opus")
+	}
+
+	channels := 1
+	if publication.TrackInfo().Stereo {
+		channels = 2
+	}
+
+	resampledPCMWriter := *writer
+	if sampleRate != opusSampleRate {
+		resampledPCMWriter = audio.ResampleWriter(*writer, sampleRate)
+	}
+
+	opusWriter, err := opus.Decode(resampledPCMWriter, channels, protoLogger.GetLogger())
+	if err != nil {
+		return nil, err
+	}
+
+	t := &OpusToPCM16AudioTrack{TrackRemote: track, opusWriter: opusWriter, pcmMWriter: resampledPCMWriter, channels: channels, sampleRate: sampleRate, logger: protoLogger.GetLogger()}
+	go t.process(handleJitter)
+	return t, nil
+}
+
+func (t *OpusToPCM16AudioTrack) process(handleJitter bool) {
+	var h rtp.Handler = rtp.NewMediaStreamIn[opus.Sample](t.opusWriter)
+	if handleJitter {
+		h = rtp.HandleJitter(int(t.TrackRemote.Codec().ClockRate), h)
+	}
+	err := rtp.HandleLoop(t.TrackRemote, h)
+	if err != nil && !errors.Is(err, io.EOF) {
+		t.logger.Errorw("error handling rtp from track", err)
+	}
+}
+
+func (t *OpusToPCM16AudioTrack) Channels() int {
+	return t.channels
+}
+
+func (t *OpusToPCM16AudioTrack) SampleRate() int {
+	return t.sampleRate
+}
+
+func (t *OpusToPCM16AudioTrack) Close() {
+	// opus writer closes resampledPCMWriter internally
+	t.opusWriter.Close()
```

Comment:
yeah, definitely harder to reader, but some comments could be helpful, like this comment is useful.

Similarly would be good to have comments on how `NewMediaStreamIn` or `opus.Decode` handle this with writer would be nice to understand how the flow actually happens

Reply:
I agree it's hard to read, mostly because of the writer APIs: https://github.com/livekit/server-sdk-go/pull/633#discussion_r2043821367

will add more comments